### ISSUE TITLE
Password based basic auth is deprecated - must switch to tokens

### ISSUE THREAD

**Issue by krisgesling** [MycroftAI/mycroft-skills-kit#32]
[Github has deprecated](https://github.blog/changelog/2019-08-08-password-based-http-basic-authentication-deprecation-and-removal/) password based http basic authentication through their API. 

Currently this sends a warning email to users when they use `msk create` but it will soon be removed entirely.

If anyone is available to help us switch to token based authentication, it would be a huge help to the project and to all Skill authors.

We propose the following flow:
1. On each run of MSK - we ask the user to generate a token
2. Provide url to https://github.com/settings/tokens/new 
3. Provide list of permissions they must select when creating the token (unclear exactly which permissions are and are not required)
4. User can store token on disk (~/.mycroft/msk?), or enter token string into MSK on each use
5. If entered into MSK token is stored in memory for that session and should never be written to disk
6. Test token is valid with `get_repos` or a similar read-only method
7. If when using the token we receive a `BadCredentialsException`, then we inform the user which permission they must add for the current functionality and link to https://github.com/settings/tokens
8. Press Enter to retry or Ctrl-C to exit.

**Comment by forslund** [comment]
I got the msk `--use-token` flag to work with a token with permissions for:
- repo (all)
- user (all)

**Comment by andlo** [comment]
As I understand the different scopes in Github Personal Access tokens one should need 
 * repo (all)

I cant see that it should need the user (all) 

The repo could be limited to only public repositories. 
 * public_repo

So either mark in repo or mark in public_repo would be enough.
![image](https://user-images.githubusercontent.com/3314671/74068016-d73d6c80-49fa-11ea-834a-98fb806664dc.png)
or
![image](https://user-images.githubusercontent.com/3314671/74068053-ec1a0000-49fa-11ea-92c0-9e855b1aab74.png)


From https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/ 


**repo:**
Grants full access to private and public repositories. That includes read/write access to code, commit statuses, repository and organization projects, invitations, collaborators, adding team memberships, deployment statuses, and repository webhooks for public and private repositories and organizations. Also grants ability to manage user projects.

**public_repo:**
Limits access to public repositories. That includes read/write access to code, commit statuses, repository projects, collaborators, and deployment statuses for public repositories and organizations. Also required for starring public repositories.

**user:**
Grants read/write access to profile info only. Note that this scope includes user:email and user:follow.

**Comment by forslund** [comment]
Public repos are probably all that's needed as you say.
I think user (probably only user:read) is currently needed for the [login check](https://github.com/MycroftAI/mycroft-skills-kit/blob/master/msk/util.py#L80). But may be able to reimplement it to do a read of some repo-information instead of the get_user() call.

**Comment by andlo** [comment]
When using msk --use-token it asks for token as expected.

But I think it should (if it could) use git credential.helper if one has stored git credentials (which do work with tokens)

I use:
```
git config --global credential.helper.store
git config --global user.name "username"
git config --global user.password "TOKEN" 
```
and then the credentials are stored in ~/.git-credentials and is used when pushing to my repos.

Same would be nice if msk could do as I defently would not remember a github token.

**Comment by andlo** [comment]
Looking into msk I think that the "fun" should be happening in 
utils.py in the function ```ask_for_github_credentials```
Changes need to be made to only use token (eg. remove the token option elsewhere in the msk)

Maybe splitout test if credentials works to a function ```check_github_credentials```, as we have to tjeck several times.

Then the flow should be:

first:
1. check if ~/.mycroft/msk/GITHUB_TOKEN exists, 
2. read token from that file 
3. check if token work - check_github_credentials()
4. keep token in memmory


Second - If no token or token dosnt work
1. ask for token and inform on how to make one
2. check if token work - check_github_credentials()
3. if works ask if user will store token
4. if yes store in  ~/.mycroft/msk/GITHUB_TOKEN
5. keep token in memmory

third continue as usaly

@forslund did mention getting the username, so maybe there are some more todo. Ill try making this happen if I can.

**Comment by forslund** [comment]
That sounds like a good plan...

Keep token in memory is not really needed to do explicitly since it should be part of the github object that's used for the interactions after the "login".

The user information is already being fetched so that's not really a TODO (apart from the current verification of the login it's used when checking if an existing PR should be updated)

**Comment by andlo** [comment]
YEP - I figured most of that out alreddy :) 
Most work is done and its looking good......

The dificult part can be how to direct user to make a token and have right scope. Think the best would be to do something like
```
          print('=== GitHub Token ===')
            print('To auhenticate to GitHub we need a GitHub token')
            print('Go to https://github.com/settings/tokens to make one')
            print('The scope for the token you oly have to check in Repo')
            print('see more on ohow to make a GitHub token for use wiht msk')
            print('in the docs on https://somepagedescribibgmskgithubscope')
```

**Comment by forslund** [comment]
I always need to have read:user for everything to work...

I think the text should be a step by step guide, something like

```
To auhenticate with GitHub a Personal Access Token is needed.

1. Go to https://github.com/settings/tokens/new create one
2. Give the token a name and Select the scopes
  - repo (which ever is the needed one/ones
  - read:user
3. Click Generate Token
4. Copy the generated token and store it somewhere safe
5. Paste it in below
```

**Comment by andlo** [comment]
It is workig :) 

But if using a token with wrong scope it dosnt chek for that :( so need a way to check that....

Should I PR the stuf I made to the master branch or should there be a token_ony branch ?

**Comment by forslund** [comment]
Just create a PR, no need for a separate branch. (core won't use it until a new release is tagged and pushed to pypi)

Currently the `github.get_use.login()` is a good check and then a check for the repos. Wonder if `github.get_user().get_repos()` is good enough?

**Comment by andlo** [comment]
That could be OK - if the user has any repos....

**Comment by forslund** [comment]
Well it should return any way just a list of zero, and if the rights aren't right it should raise an exception that the the repo list can't be read (I think)

**Comment by andlo** [comment]
Cant get that to happen.....

Made a PR https://github.com/MycroftAI/mycroft-skills-kit/pull/33

I try to check in line 94 in util.py, but if using a token without scopes it dosnt fail at that check...

**Comment by forslund** [comment]
Awesome @andlo, I'll review it tomorrow.

**Comment by andlo** [comment]
great - still missing the check for if the token is right though. But the rest seems to work. 

Thanks for letting me partisipating :)

**Comment by andlo** [comment]
Hmmmm testing for right scope could be done with github.oauth_scopes I think....But right now I cant get it to work :(

**Comment by andlo** [comment]
Figured out how....And now it works

I have testet by:
token that as scope repo
token that hassnt scope repo
token that dosnt exist

And also tested storing token
and if stored token no longer works.

**Comment by forslund** [comment]
Resolved by #33
